fraud/registry.go

```diff
@@ -9,9 +11,19 @@ var (
 	defaultUnmarshalers = map[ProofType]ProofUnmarshaler{}
 )
 
-// Register sets unmarshaler in map by provided ProofType.
-func Register(proofType ProofType, unmarshaler ProofUnmarshaler) {
+// Register adds a string representation and unmarshaller for the provided ProofType.
+func Register(p Proof) {
 	unmarshalersLk.Lock()
 	defer unmarshalersLk.Unlock()
-	defaultUnmarshalers[proofType] = unmarshaler
+	if _, ok := defaultUnmarshalers[p.Type()]; ok {
+		panic(fmt.Sprintf("fraud: unmarshaler for %s proof is registered", p.Type()))
+	}
+	defaultUnmarshalers[p.Type()] = func(data []byte) (Proof, error) {
+		// the underlying type of `p` is a pointer to a struct and assigning `p` to a new variable is not the
+		// case, because it could lead to data races.
+		// So, there is no easier way to create a hard copy of Proof other than using a reflection.
+		proof := reflect.New(reflect.ValueOf(p).Elem().Type()).Interface().(Proof)
```

Comment:
doing proof := *p would lead to a data race?

Reply:
you are not able to do  `proof := *p`, bc `p` is an interface. So, the only way is `proof := p`, but in this proof address will be copied, so getting unmarshaller from different threads will cause a data race, because you will be accessing the same address inside func.

Reply:
We are talking about different races here btw. 
@renaynay to answer yours, no, because copying is reading and reading only is race free

Reply:
I see Proof is an interface yes. Okay.

Reply:
There is a tradeoff between code readability and API simplicity. We can either pass one param as the Proof itself to Register it, or we pass two params(type and Unmarshaller) to register it. 

We consider readability being solvable if properly documented and simplifies the life of users(like you) who don't want to care or understand why the registration for unmarshalling is needed. Here, we just registered a proof and everything else is done for you